### PR TITLE
Simplify fab has access lookup

### PR DESCRIPTION
**Pull request by jhtimmins**
Simplifies the FAB security manager code by doing the following.
1. Modifies permission lookup by accessing `User.perms` property method and checking if a permission pair is included in the user's list of perms.
2. Extends `AnonymousUserMixin` to create `AnonymousUser`, which has both `perms` and `roles` property methods. Removes the need for custom auth logic for anonymous users.
3. Removes helper methods that provide little value, including:
 * `.get_readable_dags()`
 * `.current_user_has_permissions()`
 * `._has_resource_access()`
 * `.user_has_perms()`
 * `.user_has_roles()`
4. Removes unused methods:
 * `SecurityManager.is_item_public()`
5. Moves the code to create a root dag resource name to `Permissions.resource_name_for_dag()`

This is the next piece of the permissions refactor.

### CODE REVIEW

File: airflow/www/fab_security/sqla/models.py [hunk 1]
```diff
@@ -178,6 +180,13 @@ def is_active(self):
     def is_anonymous(self):
         return False
 
+    @property
+    def perms(self):
+        perms = set()
+        for role in self.roles:
+            perms.update((perm.action.name, perm.resource.name) for perm in role.permissions)
```
Comment (andrewgodwin) [inline airflow/www/fab_security/sqla/models.py:235]: I think you may need to do this as a single query for efficiency reasons?

Comment (jhtimmins) [reply airflow/www/fab_security/sqla/models.py:235]: I added `lazy=joined` to several model fields, which causes prefetching. It now runs 11 queries instead of 49.

Comment (jhtimmins) [reply airflow/www/fab_security/sqla/models.py:235]: @andrewgodwin that said, this PR does add some overall time to the tests. `pytest tests/views --with-db-init` runs in 1:53 on main and 2:00 on my branch. How significant is that?

File: airflow/api_connexion/endpoints/dag_run_endpoint.py [hunk 2]
```diff
@@ -211,7 +212,7 @@ def get_dag_runs_batch(session):
         raise BadRequest(detail=str(err.messages))
 
     appbuilder = current_app.appbuilder
-    readable_dag_ids = appbuilder.sm.get_readable_dag_ids(g.user)
+    readable_dag_ids = {dag.dag_id for dag in appbuilder.sm.get_readable_dags(g.user)}
     query = session.query(DagRun)
     if data.get("dag_ids"):
         dag_ids = set(data["dag_ids"]) & set(readable_dag_ids)
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]:
```suggestion
        dag_ids = set(data["dag_ids"]) & readable_dag_ids
```

Because `readable_dag_ids` is now already a set

Or

```suggestion
        dag_ids = readable_dag_ids.intersection(data["dag_ids"])
```

File: airflow/www/fab_security/manager.py [hunk 3]
```diff
@@ -85,6 +85,33 @@ def _oauth_tokengetter(token=None):
...
+    def roles(self, roles):
+        self._roles = roles
+        self._perms = set()
+
+    @property
+    def perms(self):
+        if not self._perms:
+            self._perms = set()
+            for role in self.roles:
+                self._perms.update({(perm.action.name, perm.resource.name) for perm in role.permissions})
```
Comment (uranusjr) [inline airflow/www/fab_security/manager.py:112]:
```suggestion
            self._perms = {
                (perm.action.name, perm.resource.name)
                for role in self.roles
                for perm in role.permissions
            }
```

Same for `User` below

Comment (jhtimmins) [reply airflow/www/fab_security/manager.py:112]: This is a fair critique, but I have a pretty strong aversion to nested comprehensions, as I'm convinced Python's design is flawed.

File: airflow/www/auth.py [hunk 4]
```diff
@@ -35,7 +35,7 @@ def decorated(*args, **kwargs):
             __tracebackhide__ = True  # Hide from pytest traceback.
 
             appbuilder = current_app.appbuilder
-            if not g.user.is_anonymous and not appbuilder.sm.current_user_has_permissions():
+            if not (g.user.is_anonymous or g.user.perms):
```
Comment (uranusjr) [inline airflow/www/auth.py]: I feel the previous version is a bit easier to understand (i.e. “not anonymous but does not have any permissions”)

```suggestion
            if not g.user.is_anonymous and not g.user.perms:
```

File: tests/www/test_security.py [hunk 5]
```diff
@@ -29,6 +29,7 @@
 from airflow.models.dag import DAG
 from airflow.security import permissions
 from airflow.www import app as application
+from airflow.www.fab_security.manager import AnonymousUser
```
Comment (ashb) [inline tests/www/test_security.py:32]: This is the only change in this file -- is it needed?

File: airflow/www/fab_security/sqla/manager.py [hunk 6]
```diff
@@ -339,7 +333,7 @@ def filter_roles_by_perm_with_action(self, action_name: str, role_ids: List[int]
             )
         ).all()
 
-    def get_role_permissions_from_db(self, role_id: int) -> List[Permission]:
+    def get_permissions_from_roles(self, role_id: int) -> List[Permission]:
```
Comment (ashb) [inline airflow/www/fab_security/sqla/manager.py]:
```suggestion
    def get_permissions_from_role(self, role_id: int) -> List[Permission]:
```
Should be singular given it takes a single role_id. (Possibly also called `_from_role_id`, but don't mind on that one)

File: airflow/www/fab_security/manager.py [hunk 7]
```diff
@@ -85,6 +85,33 @@ def _oauth_tokengetter(token=None):
...
+        return self._perms
```
Comment (ashb) [inline airflow/www/fab_security/manager.py:113]: `from airflow.compat.functools import cached_property` and then:

```suggestion
    @cached_property
    def perms(self):
        _perms = set()
        for role in self.roles:
            _perms.update({(perm.action.name, perm.resource.name) for perm in role.permissions})
        return _perms
```
And then `del self.perms` in the `@roles.setter` to clear the cache.

(This is all functionally the same, I just find the delcarative `@cached_property` clearer to read.)

File: airflow/www/auth.py (hunk 4 above)
Comment (ashb) [reply airflow/www/auth.py]: See this one @jhtimmins

File: airflow/api_connexion/parameters.py [hunk 8]
```diff
@@ -100,7 +100,7 @@ def apply_sorting(query, order_by, to_replace=None, allowed_attrs=None):
     if to_replace:
         lstriped_orderby = to_replace.get(lstriped_orderby, lstriped_orderby)
     if order_by[0] == "-":
-        order_by = f"{lstriped_orderby} desc"
+        order_by = desc(text(lstriped_orderby))
```
Comment (ashb) [inline airflow/api_connexion/parameters.py]: Unrelated change in this file?

**Review by ashb — COMMENTED** [review]
LGTM -- a few small comments to address or say "naaah" to

File: airflow/www/fab_security/manager.py (hunk 7 above)
Comment (jhtimmins) [reply airflow/www/fab_security/manager.py:113]: I'll make this change in a follow up PR

File: airflow/www/fab_security/sqla/manager.py (hunk 6 above)
Comment (jhtimmins) [reply airflow/www/fab_security/sqla/manager.py]: Its actually not used anymore so should be removed entirely

**Comment by jhtimmins** [comment]
@ashb The most recent changes address a bug that only arises with MSSQL. Bc my changes to the permission models include prefetching of data, the `parameters.apply_sorting()` method doesn't work anymore https://github.com/apache/airflow/blob/main/airflow/api_connexion/parameters.py#L99. It's a utility method for allowing users to pass sort parameters to the API, and it then adds the order_by string to the query. This is problematic for joined queries, since sqlalchemy uses aliases to reference the joins and subqueries. MSSQL has a different alias naming scheme than the other DBs.

My change addresses the two API functions that query on the permission models, and replaces the calls to `apply_sorting()` with localized code to handle the different sort arguments.

I've duplicated some code, which I could remove by modifying the original `.apply_sorting()` method, but to prevent scope creep I'd rather do that in a future PR.

File: airflow/api_connexion/endpoints/role_and_permission_endpoint.py [hunk 9]
```diff
@@ -68,11 +68,19 @@ def get_roles(*, order_by: str = "name", limit: int, offset: Optional[int] = Non
     appbuilder = current_app.appbuilder
     session = appbuilder.get_session
     total_entries = session.query(func.count(Role.id)).scalar()
+    direction = "desc" if order_by.startswith("-") else "asc"
```
Comment (ashb) [inline airflow/api_connexion/endpoints/role_and_permission_endpoint.py]: Rather than the double `getattr` we should use the standalone asc/desc functions https://docs.sqlalchemy.org/en/13/core/sqlelement.html?highlight=desc#sqlalchemy.sql.expression.desc

```python
from sqlalchemy import desc, asc
```

and then

```suggestion
    direction = desc if order_by.startswith("-") else asc
```

etc

File: airflow/api_connexion/endpoints/role_and_permission_endpoint.py [hunk 10]
```diff
@@ -68,11 +68,19 @@ def get_roles(*, order_by: str = "name", limit: int, offset: Optional[int] = Non
...
+    if order_by not in allowed_filter_attrs:
+        raise BadRequest(
+            detail=f"Ordering with '{order_by}' is disallowed or "
+            f"the attribute does not exist on the model"
+        )
+
     query = session.query(Role)
-    query = apply_sorting(query, order_by, to_replace, allowed_filter_attrs)
-    roles = query.offset(offset).limit(limit).all()
+    roles = query.order_by(getattr(getattr(Role, order_param), direction)()).offset(offset).limit(limit).all()
```
Comment (ashb) [inline airflow/api_connexion/endpoints/role_and_permission_endpoint.py]:
```suggestion
    roles = query.order_by(direction(getattr(Role, order_param)).offset(offset).limit(limit).all()
```

File: airflow/www/security.py [hunk 11]
```diff
@@ -397,62 +354,35 @@ def has_access(self, action_name, resource_name, user=None) -> bool:
...
         2. Has can_read action on dags resource.
         3. Has can_edit action on dags resource.
         """
+        if not user:
+            user = g.user
         return (
-            self._has_role(['Admin', 'Viewer', 'Op', 'User'])
-            or self._has_perm(permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG)
-            or self._has_perm(permissions.ACTION_CAN_EDIT, permissions.RESOURCE_DAG)
+            self._has_role(['Admin', 'Viewer', 'Op', 'User'], user)
```
Comment (ashb) [inline airflow/www/security.py:383]: (Unrelated to this PR, we should remove this Role check -- those roles _have_ these permissions and by having this here it means you _can't_ remove access to all dags from the stock roles.)

**Review by ashb — CHANGES_REQUESTED** [review]
Approach looks good and tidies up the SecurityManager code a lot.

One small improvement to make.

**Review by ashb — APPROVED** [review]
Made the changes from the review, and have rebased.
